src/OpenTelemetry/Internal/TagTransformer.cs

```diff
@@ -0,0 +1,138 @@
+// <copyright file="TagTransformer.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenTelemetry.Internal;
+
+internal abstract class TagTransformer<T>
+{
+    public bool TryTransformTag(KeyValuePair<string, object> tag, out T result)
+    {
+        result = default;
```

Comment:
Doesn't have to be done on this PR, but I wonder if this is expensive for large structs. Like, we initialize the value (and all its fields) and then later for happy path we override it to what we really want to return. Might be better to re-order so we only set `result = default` when we are about to bail out.